### DISCUSSION TITLE
Ver 2.0 pi4j is poor, nothing works.

### DISCUSSION THREAD

**Discussion by msawicki-poland** [Pi4J/pi4j#232] [General]
First of all, I have got original Raspberry Pi CM4 with updated Raspbian:
`Linux device 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l GNU/Linux`
I've installed pigpio to control from bash simple output GPIO21 like this:
```
pigpio m 21 w
pigpio w 21 1
```
and it's working, led's flashing.
Now, I wanted to use Pi4J (the newest) so I've installed Java:
```
openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment (build 11.0.16+8-post-Raspbian-1deb10u1)
OpenJDK Server VM (build 11.0.16+8-post-Raspbian-1deb10u1, mixed mode)
```
"Hello world" works.
Now I used sample: https://pi4j.com/documentation/io-examples/digital-output/
and create file:
```
import com.pi4j.*;
import com.pi4j.io.gpio.digital.*;

public class GettingStartedExample {

    public static void main(String[] args) throws Exception {

        var pi4j = Pi4J.newAutoContext();

        DigitalOutput output = pi4j.dout().create(21, "my-digital-output-1");
	output.high();

        pi4j.shutdown();
    }
}
```
and it's not working, Led isn't flashing.
no error during compilation and running, during running I got:
```
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
```

P.S.
1. on the page https://pi4j.com/documentation/device-examples/simpleled/ link https://github.com/Pi4J/pi4j-example-components/tree/Dev-Arcade/src/main/java/com/pi4j/example/components doesn't work
2. where can I find information about parameters of eq. pi4j.dout().create function. Docs is very poor.
3. on page https://pi4j.com/getting-started/set-up-a-new-raspberry-pi/ in the section "Install Pi4J" is "TODO" :( the key info is todo :(

**Comment by taartspi** [comment]
Pertaining to pi4j.dout not working.  See discussion 225 to create the output pin via different steps

relevant detail from 225

I believe the dout().create() creates the output pin with no operational provider. I did not see a means to add a valid provider using that pattern so I changed your code as seen below, the LED blinks and remains ON at completion.

I did not have your logging class so I modified the code to use the console. Also, for my clarification, you refer to pin 16, which is BCM16, pin 36.

            // create a digital output instance using the default digital output provider
            /*
            var output = pi4j.dout().create(DIGITAL_OUTPUT_PIN);
            output.config().shutdownState(DigitalState.HIGH);
            */
            
            DigitalOutput output = null;
            var outputConfig3 = DigitalOutput.newConfigBuilder(pi4j)
                    .address(DIGITAL_OUTPUT_PIN)
                    .shutdown(DigitalState.HIGH)
                    .provider("pigpio-digital-output");
            try {
                output = pi4j.create(outputConfig3);
            } catch (Exception e) {
                e.printStackTrace();
                console.println("create DigOut EN failed");
                System.exit(203);
            }

**Comment by msawicki-poland** [comment]
So I've changed my source code:
```
import com.pi4j.*;
import com.pi4j.io.gpio.digital.*;

public class GettingStartedExample {

    public static void main(String[] args) throws Exception {

	    var pi4j = Pi4J.newAutoContext();
            DigitalOutput output = null;
            var config = DigitalOutput.newConfigBuilder(pi4j)
                    .address(21)
                    .shutdown(DigitalState.HIGH)
                    .provider("pigpio-digital-output");
            try {
                output = pi4j.create(config);
            } catch (Exception e) {
		System.out.println("create output failed");
            }
    }
}

```
and try to run as a root:
```
# export CLASSPATH=$CLASSPATH:/opt/pi4j/lib/*
# javac GettingStartedExample.java 
# pigpiod
# pigs m 21 w
# pigs w 21 1
# pigs w 21 0
# java GettingStartedExample
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
[main] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
create output failed
```
and it's not working again. Look that pigpio as bash tool and deamon works but from Java I got "**PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed"** and the exception was thrown.

**Reply by taartspi** [reply]
I think your comment about ‘root’ is in the correct direction as likely your failure is authority.  I accomplish the needed authority by using the  ‘sudo’ command to assume superuser

**Reply by msawicki-poland** [reply]
```
$ sudo java GettingStartedExample
Exception in thread "main" java.lang.NoClassDefFoundError: com/pi4j/Pi4J
	at GettingStartedExample.main(GettingStartedExample.java:8)
Caused by: java.lang.ClassNotFoundException: com.pi4j.Pi4J
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 1 more

```

**Comment by taartspi** [comment]
I used the Minimal example steps to set up the environment.

mkdir Example
pi@raspi464alt:~ $ cd Example/
pi@raspi464alt:~/Example $ git clone https://github.com/Pi4J/pi4j-example-minim


I updated the pom file. I think I had problems with 2.1.1. and moved past it some time back. My changes were


       <pi4j.version>2.2.0-SNAPSHOT</pi4j.version>


        <repository>
            <id>oss-snapshots-M2-repo</id>
            <name>Sonatype OSS Maven2 Repository</name>
            <url>https://mvnrepository.com/artifact/org.apache.felix/maven-bundle-plugin</url>
        </repository>
        <repository>
            <id>sonatype-oss-staging</id>
            <name>Sonatype Open Source Maven Repository</name>
            <url>https://oss.sonatype.org/service/local/staging/deploy/maven2</url>
        </repository>






cd to
/home/pi/Example/pi4j-example-minimal/src/main/java/com/pi4j/example
GettingStartedExample.java

package com.pi4j.example;

import com.pi4j.*;
import com.pi4j.io.gpio.digital.*;

public class GettingStartedExample {

    public static void main(String[] args) throws Exception {

        var pi4j = Pi4J.newAutoContext();
        DigitalOutput output = null;
        var config = DigitalOutput.newConfigBuilder(pi4j)
                .address(21)
                .shutdown(DigitalState.HIGH)
                .provider("pigpio-digital-output");
        try {
            output = pi4j.create(config);
        } catch (Exception e) {
            System.out.println("create output failed");
        }
     System.out.println("Hello  !!!");
    }
}






cd ~/Example/pi4j-example-minimal
mvn clean package



cd to classes
cd /home/pi/Example/pi4j-example-minimal/target/classes

 
sudo java -classpath /home/pi/Example/pi4j-example-minimal/target/distribution/*:/home/pi/Example/pi4j-example-minimal/target/classes   com.pi4j.example.GettingStartedExample



~/Example/pi4j-example-minimal/target/classes $ 
sudo java -classpath /home/pi/Example/pi4j-example-minimal/target/distribution/*:/home/pi/Example/pi4j-example-minimal/target/classes   com.pi4j.example.GettingStartedExample
/home/pi/Example/pi4j-example-minimal/target/classes
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
Hello  !!!




Tom
